src/IRMatch.h

```diff
@@ -45,6 +47,1366 @@ bool expr_match(Expr pattern, Expr expr, std::map<std::string, Expr> &result);
+struct Const {
+    struct pattern_tag {};
+    HALIDE_ALWAYS_INLINE bool match(const BaseExprNode &e, MatcherState &state) const {
+        const BaseExprNode *op = &e;
+        if (e.node_type == IRNodeType::Broadcast) {
+            op = ((const Broadcast *)op)->value.get();
+        }
+        switch (op->node_type) {
+        case IRNodeType::IntImm:
+            return ((const IntImm *)op)->value == i;
+        case IRNodeType::UIntImm:
+            return ((const UIntImm *)op)->value == i;
+        case IRNodeType::FloatImm:
+            return ((const FloatImm *)op)->value == i;
+        default:
+            return false;
+        }
+    }
+
+    HALIDE_ALWAYS_INLINE bool match(const Const<i> &, MatcherState &state) const {
+        return true;
+    }
+};
+
+template<int i>
+std::ostream &operator<<(std::ostream &s, const Const<i> &op) {
+    s << i;
+    return s;
+}
+
+template<typename Op>
+int64_t constant_fold_bin_op(halide_type_t, int64_t, int64_t, uint32_t &);
+
+template<typename Op>
+uint64_t constant_fold_bin_op(halide_type_t, uint64_t, uint64_t, uint32_t &);
+
+template<typename Op>
+double constant_fold_bin_op(halide_type_t, double, double, uint32_t &);
+
+template<typename Op>
+uint64_t constant_fold_cmp_op(halide_type_t, int64_t, int64_t);
+
+template<typename Op>
+uint64_t constant_fold_cmp_op(halide_type_t, uint64_t, uint64_t);
+
+template<typename Op>
+uint64_t constant_fold_cmp_op(halide_type_t, double, double);
+
+
+// Matches one of the binary operators
+template<typename Op, typename A, typename B>
+struct BinOp {
+    struct pattern_tag {};
+    A a;
+    B b;
+    HALIDE_ALWAYS_INLINE bool match(const BaseExprNode &e, MatcherState &state) const {
+        if (e.node_type != Op::_node_type) {
+            return false;
+        }
+        const Op &op = (const Op &)e;
+        return a.match(*op.a.get(), state) && b.match(*op.b.get(), state);
+    }
+
+    template<typename Op2, typename A2, typename B2>
+    HALIDE_ALWAYS_INLINE bool match(const BinOp<Op2, A2, B2> &op, MatcherState &state) const {
+        return std::is_same<Op, Op2>::value && a.match(op.a, state) && b.match(op.b, state);
+    }
+
+    HALIDE_ALWAYS_INLINE Expr make(MatcherState &state) const {
+        Expr ea = to_expr(a, state), eb = to_expr(b, state);
+        match_types(ea, eb);
+        return Op::make(std::move(ea), std::move(eb));
+    }
+
+    template<typename A1 = A,
+             typename B1 = B,
+             typename = decltype(std::declval<A1>().make_folded_const(std::declval<MatcherState &>())),
+             typename = decltype(std::declval<B1>().make_folded_const(std::declval<MatcherState &>()))>
+    HALIDE_ALWAYS_INLINE MatcherState::Const make_folded_const(MatcherState &state) const {
+        MatcherState::Const ca = a.make_folded_const(state);
+        MatcherState::Const cb = b.make_folded_const(state);
+        MatcherState::Const result;
+        result.type = ca.type;
+        result.flags = ca.flags | cb.flags;
```

Comment:
Is it actually possible to have both signed_integer_overflow and indeterminate_expression set on the same result?

Reply:
I think if you have an integer overflow plus a divide by zero you'll get something with both set. The constant folder currently just picks one of the two arbitrarily. Any preference for which should dominate?

Reply:
Not sure. I'd guess 'indeterminate' since that is sorta-more-fatal but I dunno without some examination of an actual use case.